mesonbuild/minstall.py

```diff
@@ -741,7 +765,41 @@ def rebuild_all(wd: str) -> bool:
         print("Can't find ninja, can't rebuild test.")
         return False
 
-    ret = subprocess.run(ninja + ['-C', wd]).returncode
+    def drop_privileges() -> T.Tuple[T.Optional[EnvironOrDict], T.Optional[T.Callable[[], None]]]:
+        if os.geteuid() == 0:
+            import pwd
+            env = os.environ.copy()
+
+            if os.environ.get('SUDO_USER') is not None:
+                orig_user = env.pop('SUDO_USER')
+                orig_uid = env.pop('SUDO_UID', 0)
+                orig_gid = env.pop('SUDO_GID', 0)
+                homedir = pwd.getpwuid(int(orig_uid)).pw_dir
+            elif os.environ.get('DOAS_USER') is not None:
+                orig_user = env.pop('DOAS_USER')
+                pwdata = pwd.getpwnam(orig_user)
+                orig_uid = pwdata.pw_uid
+                orig_gid = pwdata.pw_gid
+                homedir = pwdata.pw_dir
+            else:
+                return None, None
+
+            env['USER'] = orig_user
+            env['HOME'] = homedir
+
+            def wrapped() -> None:
+                print(f'Dropping privileges to {orig_user!r} before running ninja...')
+                if orig_gid is not None:
+                    os.setgid(int(orig_gid))
+                if orig_uid is not None:
+                    os.setuid(int(orig_uid))
+
+            return env, wrapped
+        else:
+            return None, None
+
+    env, preexec_fn = drop_privileges()
+    ret = subprocess.run(ninja + ['-C', wd], env=env, preexec_fn=preexec_fn).returncode
```

Comment:
In the case drop_privileges() returns None, or preexec_fn is not supported (posix only), should we abort with an error instead?

Reply:
This is really only supposed to handle posix in the first place as that's where we might need to drop permissions. In theory this should never return geteuid() is 0 on Windows?

The other case where this can return None is when it's not running under either sudo or doas. In such a case, the login user might be root and all work is being performed as root, so we shouldn't error out. It's fine to compile as root if you downloaded sources as root and ran `setup` as root.

So I think this should stay as-is.

Reply:
> It's fine to compile as root if you downloaded sources as root and ran setup as root.

That's a good point, and will almost always be the case in docker/CI cases actually.

Reply:
More generally, we should return None here on Windows because geteuid cannot return 0, it doesn't exist. :D